qiskit/primitives/backend_estimator_v2.py

```diff
@@ -160,9 +160,15 @@ def _validate_pubs(self, pubs: list[EstimatorPub]):
 
     def _run(self, pubs: list[EstimatorPub]) -> PrimitiveResult[PubResult]:
         pub_dict = defaultdict(list)
+        for pub in pubs:
+            data = self._preprocess_pub(pub)
+        bc_obs = data.observables
+        coeff_sq_sum = 0
+        for pauli, coeff in bc_obs[0].items():
+            coeff_sq_sum += coeff**2
         # consolidate pubs with the same number of shots
         for i, pub in enumerate(pubs):
-            shots = int(math.ceil(1.0 / pub.precision**2))
+            shots = int(math.ceil(coeff_sq_sum / pub.precision**2))
```

Comment:
Why not make coeff_sq_sum depend on the pub? With what you've written, it only depends on the last pub.

Moreover, it only depends on the first element of the last pub.

Reply:
Sounds good. I guess it was indirectly dependent it. 

From the EstimatorV2 tutorial example bc_obs is of the form [{'II': 1.0, 'IZ': 2.0, 'XI': 3.0}] so bc_obs[0] gets all the dictionary terms {'II': 1.0, 'IZ': 2.0, 'XI': 3.0}.